File: JavaBaseWrapper/src/main/java/eu/fau/cs7/daceDS/Component/Producer.java
*/*
MIT License

Copyright 2021 Moritz Gütlein

This code was originally published under the Apache 2.0 License (http://www.apache.org/licenses/LICENSE-2.0) in 2021.
In 2025, it has been relicensed under the MIT License (https://choosealicense.com/licenses/mit/) with the explicit permission of all copyright holders.
*/
package eu.fau.cs7.daceDS.Component;

import java.util.HashMap;
import java.util.List;


/**
 * A generic producer class.
 *  
 * @author guetlein
 *
 * @param <T>
 */
public abstract class  Producer<T>
{

	protected long time = -1;
	protected static long step = 1;
	protected String id;
    
    protected static HashMap<String, Long> sentCounting = new HashMap<String, Long>();
	protected boolean countingProducer = false;
	protected boolean exactlyOnce = false; 
	protected boolean plainStringValue = true; 


	public Producer(String id){
		this.id = id;
	}

	public Producer(String id, boolean counting){
		this.id = id;
		this.countingProducer = counting;
	}

	public void setExactlyOnce(boolean b){
		this.exactlyOnce = b;
	}


	public abstract void init();
	
	public abstract void initPlainString();

	public abstract void initTopicNamingStrategy();

	public abstract void createTopics(List<String> topicNames);

	public abstract boolean publish(String topic, T payload, long time);

	public abstract boolean publish(String topic, String key, T payload, long time, int epoch, String sender);
	
	public static HashMap<String, Long> getSentCounting(){
		return sentCounting;
	}

	public abstract void close();
}

File: CppBaseWrapper/src/communication/Producer.cpp
/*
MIT License

Copyright 2021 Moritz Gütlein

This code was originally published under the Apache 2.0 License (http://www.apache.org/licenses/LICENSE-2.0) in 2021.
In 2025, it has been relicensed under the MIT License (https://choosealicense.com/licenses/mit/) with the explicit permission of all copyright holders.
*/
#include "Producer.h"
using namespace daceDS;

std::map<std::string, int64_t> Producer::countingSentMessages;

bool Producer::publish(std::string topic, std::string payload) {
    return false;
}

bool Producer::publish(std::string topic, std::string payload, int64_t time) {
    return false;
}

bool Producer::publish(std::string topic, std::string key, std::string payload, int64_t time, bool repeatOnError) {
    if (!isCounting()) {
        KDBGCB("Published to " << topic << " --> not counting");
        return false;
    }

    KDBGCB("Published to " << topic << " --> counting");
    if (Producer::countingSentMessages.count(topic) == 0) {
        Producer::countingSentMessages[topic] = 0;
    }
    Producer::countingSentMessages[topic]++;
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload, bool repeatOnError) {
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload) {
    return false;
}

bool Producer::publish(std::string topic, std::vector<char>& payload, int64_t time) {
    return false;
}

bool Producer::publish(std::string topic, std::string key, std::vector<char>& payload, int64_t time, bool repeatOnError) {
    if (!isCounting()) {
        KDBGCB("Published to " << topic << " --> not counting");
        return false;
    }

    KDBGCB("Published to " << topic << " --> counting");
    if (Producer::countingSentMessages.count(topic) == 0) {
        Producer::countingSentMessages[topic] = 0;
    }
    Producer::countingSentMessages[topic]++;
    return false;
}
